### DISCUSSION TITLE
slow write when bluefs_buffered_io=true and buffer memory is close to osd memory

### DISCUSSION THREAD

**Discussion by caffeinism** [rook/rook#10272] [Q&A]
Hello again! Back at https://github.com/rook/rook/discussions/10198.

Even after I wrote that article, I worked tirelessly to find the cause for two weeks. I found a piecemeal fact that it has something to do with osd memory size. I made a custom metric exporter to observe the thread load and the bluestore mempool as well, but I felt I couldn't find the problem. However, while observing the host's htop, I noticed that the cluster slowed down dramatically when the buffer memory became close to the memory limit of the osd spec.

It is true that it is periodically slowed down by rocksdb compaction, but I don't think that's the main cause.

![image](https://user-images.githubusercontent.com/39237502/168561518-e15a4fdc-83e4-4fcb-9aba-f68f31a20412.png)

I didn't think that the buffer memory was a big problem because the kernel's buffer memory can always be freed. How did the container's memory limit work badly? Also, I wonder if there is an alternative to this.

**Comment by caffeinism** [comment]
![image](https://user-images.githubusercontent.com/39237502/168571171-fe1ba8d0-f784-4bcc-a743-0210d2192186.png)
This is a performance chart of rocksdb allocated to ramdisk to exclude performance indicators of metadata devices.
Measurements were run at 1 OSD to eliminate other influences.

**Reply by caffeinism** [reply]
Also, even if ramdisk is used as a metadata device, if bluefs_buffered_io is set to true, the same low performance as in the first picture of the question.

**Comment by caffeinism** [comment]
Also, if the container's limit is not set, no problem seems to occur even if bluefs_buffered_io is enabled.

But I think containers should always have appropriate resource limits. I also prefer requests=limits to guarantee performance.

Is there any way to solve this?

I'm using ceph 17.2, rook 1.9, ubuntu 22.04, kubernetes 1.22.8 and containerd 1.5.8.

**Reply by travisn** [reply]
Agreed, the memory requests/limits are recommended to keep the osd memory at a predictable level. 
@markhpc Since you commented on the original issue, any thoughts on the perf issue here?

**Reply by mheler** [reply]
@caffeinism there was a discussion recently on the ceph-users mailing list in regards to this subject under low memory conditions
https://www.mail-archive.com/ceph-users@ceph.io/msg30971.html

Then additionally here in a recent PR being proposed
https://github.com/ceph/ceph/pull/65009#issuecomment-3210941667
